src/funk/fd_funk_txn.h

```diff
@@ -0,0 +1,404 @@
+   frozen (i.e. has children) and 0 otherwise (i.e. has no children).
+   Assumes txn points to an in-preparation transaction in the caller's
+   address space. */
+
+FD_FN_PURE static inline int
+fd_funk_txn_is_frozen( fd_funk_txn_t const * txn ) {
+  return !fd_funk_txn_idx_is_null( fd_funk_txn_idx( txn->child_head_cidx ) );
+}
+
+/* fd_funk_txn_is_only_child returns 1 if the in-preparation transaction
+   txn if is any only child and 0 if it has one or more siblings.
+   Assumes txn points to an in-preparation transaction in the caller's
+   address space. */
+
+FD_FN_PURE static inline int
+fd_funk_txn_is_only_child( fd_funk_txn_t * txn ) {
+  return ( fd_funk_txn_idx_is_null( fd_funk_txn_idx( txn->sibling_prev_cidx ) ) ) &
+         ( fd_funk_txn_idx_is_null( fd_funk_txn_idx( txn->sibling_next_cidx ) ) );
+}
+
+/* Operations */
+
+/* fd_funk_txn_ancestor returns a pointer in the caller's address space
+   to the youngest transaction among in-preparation transaction txn and
+   its ancestors that currently has siblings.  Returns NULL if all
+   transactions back to the root transaction have no siblings (e.g.
+   there are no competing transaction histories and thus publishing
+   transaction txn will not require canceling any other competing
+   transactions).  This is a reasonably fast O(length of ancestor
+   history) time (theoretical minimum) and a reasonably small O(1) space
+   (theoretical minimum).  This is not fortified against transaction map
+   data corruption. 
+
+   fd_funk_txn_descendant returns a pointer in the caller's address
+   space to the first the first transaction among txn and its youngest
+   direct descendant inclusive that currently either has no children or
+   has multiple children.  Returns NULL if txn is not an only child.
+   This is a reasonably fast O(length of descendant history) time
+   (theoretical minimum) and a reasonably small O(1) space (theoretical
+   minimum).  This is not fortified against transaction map data
+   corruption. 
+
+   That is, if txn's ancestor is NULL, all transactions up to and
+   including txn's descendant (which will be non-NULL) can be published
+   without cancelling any competing transactions.  Further, if the
+   descendant has a child, it has multiple children.  And if has no
+   children, all transactions in preparation are linear from the root to
+   the descendant.
+
+   In code:
+
+     if( !fd_funk_txn_ancestor( txn, map ) )
+       fd_funk_publish( funk, fd_funk_txn_descendant( funk, map ) );
+
+   would publish as much currently uncontested transaction history as
+   possible around txn.
+
+   Assumes map == fd_funk_txn_map( funk, fd_funk_wksp( funk ) ), funk is
+   a current local join and txn points to an in-preparation transaction
+   of funk in the caller's address space.  The lifetime of the returned
+   pointer is as described in fd_funk_txn_query. */
+
+FD_FN_PURE static inline fd_funk_txn_t *
+fd_funk_txn_ancestor( fd_funk_txn_t * txn,
```

Comment:
Are ancestor and descendant used for anything? Not sure I see the need for these

Reply:
Not yet.  These are just realizations that seemed useful of concepts that were talked about in the documentation (e.g. prepare / cancel / publish talk a lot about ancestors, descendants, uncontested histories, etc).  Seemed like it might be useful for generic fork management to give the user the ability to ask questions like "will publishing this cancel any competing transaction histories" or hopefully help user deal with lots of intermediate transactions if there are long unresolved forks.  I have no particular love of these functions though.